nilearn/masking.py

```diff
@@ -665,7 +665,15 @@ def _unmask_nd(X, mask, order="C"):
     if mask.dtype != np.bool:
         raise ValueError("mask must be a boolean array")
     if X.ndim != 2:
-        raise ValueError("X must be a 2-dimensional array")
+        raise TypeError("X must be a 2-dimensional array")
+    n_features = mask.sum()
+    if X.shape[-1] != n_features:
+        # Handle (potential) transpose as a special case, as this
+        #   mistake can lead to system-fatal memory allocation below.
+        if X.shape[0] == n_features:
+            X = X.T
+        else:
+            raise TypeError('X must be of shape (samples, %d).' % n_features)
```

Comment:
+1

Sent from my phone. Please forgive brevity and mis spelling



On Feb 24, 2015, 11:00, at 11:00, "Loïc Estève" <notifications@github.com> wrote:
>> @@ -665,7 +665,15 @@ def _unmask_nd(X, mask, order="C"):
>>      if mask.dtype != np.bool:
>>          raise ValueError("mask must be a boolean array")
>>      if X.ndim != 2:
>> -        raise ValueError("X must be a 2-dimensional array")
>> +        raise TypeError("X must be a 2-dimensional array")
>> +    n_features = mask.sum()
>> +    if X.shape[-1] != n_features:
>> +        # Handle (potential) transpose as a special case, as this
>> +        #   mistake can lead to system-fatal memory allocation
>below.
>> +        if X.shape[0] == n_features:
>> +            X = X.T
>> +        else:
>> +            raise TypeError('X must be of shape (samples, %d).' %
>n_features)
>
>I'd be enclined to always throw if `X.shape[-1] != n_features` rather
>than have a special edge case that automagically works.
>
>
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/nilearn/nilearn/pull/459/files#r25240484